src/java/org/apache/cassandra/index/sai/plan/Operation.java

```diff
@@ -0,0 +1,335 @@
+                        // map entries
+                        isMultiExpression = indexContext.isNonFrozenCollection();
+                        break;
+
+                    case CONTAINS:
+                    case CONTAINS_KEY:
+                        isMultiExpression = true;
+                        break;
+                }
+                if (isMultiExpression)
+                {
+                    while (analyzer.hasNext())
+                    {
+                        final ByteBuffer token = analyzer.next();
+                        perColumn.add(new Expression(indexContext).add(e.operator(), token.duplicate()));
+                    }
+                }
+                else
+                // "range" or not-equals operator, combines both bounds together into the single expression,
+                // if operation of the group is AND, otherwise we are forced to create separate expressions,
+                // not-equals is combined with the range iff operator is AND.
+                {
+                    Expression range;
+                    if (perColumn.size() == 0 || op != OperationType.AND)
+                    {
+                        range = new Expression(indexContext);
+                        perColumn.add(range);
+                    }
+                    else
+                    {
+                        range = Iterables.getLast(perColumn);
+                    }
+
+                    if (!TypeUtil.isLiteral(indexContext.getValidator()))
+                    {
+                        range.add(e.operator(), e.getIndexValue().duplicate());
+                    }
+                    else
+                    {
+                        while (analyzer.hasNext())
+                        {
+                            ByteBuffer term = analyzer.next();
+                            range.add(e.operator(), term.duplicate());
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                analyzer.end();
+            }
+        }
+
+        return analyzed;
+    }
+
+    private static int getPriority(Operator op)
+    {
+        switch (op)
+        {
+            case EQ:
+            case CONTAINS:
+            case CONTAINS_KEY:
+                return 5;
+
+            case GTE:
+            case GT:
+                return 3;
+
+            case LTE:
+            case LT:
+                return 2;
+
+            default:
+                return 0;
+        }
+    }
+
+    static RangeIterator buildIterator(QueryController controller)
+    {
+        return Node.buildTree(controller.filterOperation()).analyzeTree(controller).rangeIterator(controller);
+    }
+
+    static FilterTree buildFilter(QueryController controller)
+    {
+        return Node.buildTree(controller.filterOperation()).buildFilter(controller);
+    }
+
+    public static abstract class Node
```

Comment:
nit: Let's reduce visibility on the `Node` hierarchy as much as we can, even if it's just package-private/default.

Reply:
Reduced to package protected to allow the tests to get at them